c/tskit/trees.c

```diff
@@ -8516,3 +8539,79 @@ tsk_treeseq_pair_coalescence_counts(const tsk_treeseq_t *self,
         sample_sets, num_set_indexes, set_indexes, num_windows, windows, num_bins,
         node_bin_map, pair_coalescence_weights, num_bins, NULL, options, result);
 }
+
+static int
+pair_coalescence_quantiles(tsk_size_t input_dim, const double *weight,
+    const double *values, tsk_size_t output_dim, double *output, void *params)
+{
+    int ret = 0;
+    double coalesced, timepoint;
+    double *quantiles = (double *) params;
+    tsk_size_t i, j;
+    j = 0;
+    coalesced = 0.0;
+    timepoint = -INFINITY;
+    for (i = 0; i < input_dim; i++) {
+        if (weight[i] > 0) {
+            coalesced += weight[i];
+            if (values[i] <= timepoint) {
+                ret = TSK_ERR_UNSORTED_TIMES;
+                goto out;
+            }
+            timepoint = values[i];
+            while (j < output_dim && quantiles[j] <= coalesced) {
+                output[j] = timepoint;
+                j += 1;
+            }
+        }
+    }
+    if (quantiles[output_dim - 1] == 1.0) {
+        output[output_dim - 1] = timepoint;
+    }
+out:
+    return ret;
+}
+
+static int
+check_quantiles(const tsk_size_t num_quantiles, const double *quantiles)
+{
+    int ret = 0;
+    tsk_size_t i;
+    double last = -INFINITY;
+    for (i = 0; i < num_quantiles; ++i) {
+        if (quantiles[i] <= last || quantiles[i] < 0.0 || quantiles[i] > 1.0) {
+            ret = TSK_ERR_BAD_QUANTILES;
+            goto out;
+        }
+        last = quantiles[i];
+    }
+out:
+    return ret;
+}
+
+int
+tsk_treeseq_pair_coalescence_quantiles(const tsk_treeseq_t *self,
+    tsk_size_t num_sample_sets, const tsk_size_t *sample_set_sizes,
+    const tsk_id_t *sample_sets, tsk_size_t num_set_indexes, const tsk_id_t *set_indexes,
+    tsk_size_t num_windows, const double *windows, tsk_size_t num_bins,
+    const tsk_id_t *node_bin_map, tsk_size_t num_quantiles, double *quantiles,
+    tsk_flags_t options, double *result)
+{
+    int ret = 0;
+    void *params;
```

Comment:
```suggestion
    void *params = (void *) quantiles;
```

saves a line (and a bit of human parsing)